### PR TITLE
Add option to enable/disable override of raycast

### PR DESCRIPTION
**Pull request by liberostelios**
A potential solution for #168

### CODE REVIEW

**Comment by gkjohnson** [comment]
Thanks for making the PR!

I think I'd like to avoid traversing over all tiles when modifying the `overrideRaycast` field so there's no surprise performance overhead when toggling the field. It can also be a bit tricky to follow because new tiles are often added and removed from the hierarchy as the camera moves around.

I was imagining keeping `TilesRenderer.raycast` unmodified and instead adjusting [TileGroup.raycast](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/7ad0e30d2c9eaf352842a502678835b720452bb7/src/three/TilesGroup.js#L16-L20) to check if `tilesRenderer.optimizeRaycast` was true before calling `TilesRenderer.raycast`:

```js
raycast( ... ) {

  if ( this.tilesRenderer.optimizeRaycast ) {
    
    this.tilesRenderer.raycast( ... )

  }

}
```

And instead of `emptyRaycast` we can replace the built raycast on all loaded tiles meshes [here](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/7ad0e30d2c9eaf352842a502678835b720452bb7/src/three/TilesRenderer.js#L617-L622) with one that checks whether optimizedRaycast is false:

```js
function overriddenRaycast( ... ) {

  if ( ! tilesRenderer.optimizeRaycast ) {

    Object.getPrototypeOf( this ).raycast.call( this, ... );

  }

}
```

That seems like it would have fewer moving parts to me. What do you think?

**Comment by liberostelios** [comment]
That's right. I haven't thought of using `emptyRaycast` as a proxy to call the original `raycast` function. I'll do that.

**Comment by liberostelios** [comment]
I've implemented the aforementioned idea. The only thing missing was that `overridenRaycast()` didn't have access to the tilesRenderer so I had to add this to each object's `userData`.

I also realized that the previous option in the example wasn't work.

I tried to verify that this works by some custom console log. I don't know if you have a better way to verify that it works.

File: src/three/TilesRenderer.js
Comment (gkjohnson) [inline src/three/TilesRenderer.js:39]: Great! This is what I was thinking. However instead of putting the tiles renderer in userdata lets create a new function per tiles renderer instead that can be used for the objects (from what I can tell userdata is really intended for serializable data). So the function can be created in the TilesRenderer constructor:

```js
constructor( ... ) {

  // ...

  const tilesRenderer = this;
  this._overridenRaycast = function( raycaster, intersects ) {

    if ( tilesRenderer.optimizeRaycast ) {

      Object.getPrototypeOf( this ).raycast.call( this, raycaster, intersects );

    }

  };

}
```

And then we can use that function on each of the loaded objects:

```js
c.raycast = this._overridenRaycast;
```

File: src/three/TilesRenderer.js
Comment (gkjohnson) [inline src/three/TilesRenderer.js:190]: I don't think we should have to adjust the `TilesRenderer.raycast` function. Instead we should adjust [TilesGroup.raycast](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/master/src/three/TilesGroup.js#L16-L20) to check if the tiles renderer (which it already has a reference to) has the tiles override flag is set to true. That way `TilesRenderer.raycast` will always perform the optimized version. Perhaps we can rename this function and expose it in a future release so it can be usable independent of three.js' Raycaster mechanism.

**Review by gkjohnson — CHANGES_REQUESTED** [review]
Awesome, thank you! Made a couple comments above. This is more along the lines of what I was thinking.

> I tried to verify that this works by some custom console log. I don't know if you have a better way to verify that it works.

I think adjusting the kitchen sink example (`example/index.js`) to get an array of leaf nodes of the tiles to raycast against, which is what it sounded like you were doing in your app, would be a good way to enable testing of the feature. That way it's clear if it breaks in the future.

Last thing would be to update the docs to document the new option!

File: src/three/TilesRenderer.js
Comment (liberostelios) [reply src/three/TilesRenderer.js]: I think this is resolved now.

File: src/three/TilesRenderer.js
Comment (liberostelios) [reply src/three/TilesRenderer.js]: Also resolved.

**Comment by liberostelios** [comment]
I think the latest commit must have improved the way things work. I'll write the docs as well.

I, also, locally changed the example to do an `intersectObjects()` on the meshes themselves, instead of a recursive raycast on the main group. I chose not to push this, though. Do you think it's worth having it there? We could have an extra option (like "PURE_THREEJS" in the params of the example and in this case use the different logic to make sure we just intersect the meshes. Not sure if I see a benefit on that, but it's an easy change and I have zero problems doing this.

**Comment by gkjohnson** [comment]
@liberostelios 

The code looks great now thanks for the changes! I'll wait until the docs are updated to merge.

> Do you think it's worth having it there? We could have an extra option (like "PURE_THREEJS" in the params of the example and in this case use the different logic to make sure we just intersect the meshes.

I think the code is easy enough to follow and validate. If this breaks some time in the future and we need a better way to test we can reconsider. Automated tests would be good for a case like this but setting that up for three.js and the tiles renderer is a task for another day, I think.

**Comment by gkjohnson** [comment]
@liberostelios Looks like the docs have been updated. Is this all ready to go? If so I'll give it all another look and we can get it merged!

**Comment by liberostelios** [comment]
Sorry, my bad. Forgot to mention it here. Yes, I am finished with the docs. I hope that's good enough.

**Comment by gkjohnson** [comment]
Great! Thanks again for adding this feature and getting everything updated. It all looks good. I'll publish a new version in the next week or so.

**Comment by gkjohnson** [comment]
Just published in v0.2.10. Thanks again!

**Pull request merged**
